### ISSUE TITLE
Add opt-out for evilification of modes that don't edit text

### ISSUE THREAD

**Issue by mbertheau** [syl20bnr/spacemacs#3022]
Modes like a lot of the magit modes, gnus and probably more like mu4e don't edit text directly and have their own modal paradigm. Usually this paradigm and its key bindings are documented with the package. 

Evilifying, while consistent and learnable, still breaks these docs for new users. Sometimes there are even on-screen key binding docs provided by these modes, and they don't work because the mode has been evilified.

The idea is to not evilify modes that have their own modal paradigm. Since this would be a pretty radical change in direction for spacemacs, I expect that evilification would have to be an opt-out.

Inspired by https://github.com/syl20bnr/spacemacs/issues/2490#issuecomment-126804760

**Comment by syl20bnr** [comment]
This is at the core of the Spacemacs UI, I doubt we can provide other support that setting your `dotspacemacs-editing-style` to `emacs`.
Evilifying follows a logical rebinding of commands, once you know the logic you can follow the official manual and translate the evilified key bindings.
See https://github.com/syl20bnr/spacemacs/blob/develop/doc/CONVENTIONS.org#evilify-buffers

**Comment by syl20bnr** [comment]
Evilified buffers deserve their own section in the documentation. Conventions could be made clearer.

**Comment by syl20bnr** [comment]
In a nutshell, keys are rebound following this ordering (example with `h`):
`h` > `H` > `C-h` > `C-H`
If there is no free key binidngs the macro will emit a warning in the Messages buffer.

**Comment by syl20bnr** [comment]
Spam :-)
Examples with `discard` in magit which is originally on `k`, in evilified state it is `K`, refresh is on `C-G` etc (`g` and `G` are taken and `C-g` is a funbamental emacs key bindings, so it is rebound to `C-G`).

**Comment by mbertheau** [comment]
@syl20bnr I know all that - it seems I wasn't very good at getting across the point of this issue :)

Vim is about modal editing as we all know, and it has normal mode that has many useful commands on modifier-less, often alphabetical-key bindings.

Now spacemacs carries over most of the key bindings from vim normal mode. These include commands for moving the cursor/point and editing among others. The key point is that these bindings are without modifier.

Now some modes, in vim and emacs, aren't really for editing text but for looking at a git log, or reading your email. Acknowledging that, mode authors bound modifier-less alphabetical keys to commands in a way that forms a complete paradigm for the mode at hand.

Evilifying a mod, while – as you described – deterministic, changes that paradigm, in some cases (#2490) to the point that important commands go without a key binding.

The idea is to let the user opt out of evilifying non-editing modes that have their own modifier-less alphabetical key bindings that are consistent with on-screen and external documentation.

**Comment by sooheon** [comment]
I agree with @mbertheau here. Think about it like offering editing styles of vim, emacs, and hybrid--we should also offer normal or navigation style of evilified and stock emacs.

**Comment by justbur** [comment]
@mbertheau Maybe you can just configure evil so that it starts in emacs state in the buffers you don't want evilified? Then you don't have to worry about messing with the evilified keys

**Comment by justbur** [comment]
I just tried this and it seems to work
```elisp
(add-to-list 'evil-emacs-state-modes 'dired-mode)
```

**Comment by bmag** [comment]
Just to be clear, opting out of evilified state just means moving a mode from `evil-evilified-state-modes` to `evil-emacs-state-modes`, right? I can think of three solutions for this:

1. Provide two functions that users can put in `dotspacemacs/user-config`: `(spacemacs/cancel-mode-evilifaction MODE)` to cancel evilifcation of a mode by moving it from `evil-evilified-state-modes` to `evil-emacs-state-modes`, and `(spacemacs/cancel-all-evilification)` to cancel evilification of all evilified mode.

2. Provide a `dotspacemacs-non-evilified-modes` variable, which will be respected by the `evilify` macro. `evilify` won't evilify modes in `dotspacemacs-non-evilified-modes`. A special value of `all` will mean to never evilify a mode.

3. Let the user do it manually in `dotspacemacs/user-config`, as shown by @justbur 

Personally I prefer (2), because it avoids potential lazy-loading issues.

**Comment by justbur** [comment]
@bmag It looks like [this is the function](https://github.com/syl20bnr/spacemacs/blob/develop/core/core-evilified-state.el#L154-L159) that you could change to address 2

**Comment by syl20bnr** [comment]
Just a quick side note on this (because I don't agree).
> Now spacemacs carries over most of the key bindings from vim normal mode. These include commands for moving the cursor/point and editing among others. The key point is that these bindings are without modifier.

I use magit extensively everyday and I have no issue with the few modifiers introduced because they are not often used, the comfort brought by hjkl and other keys in evilified state is far more important. Consistency is also upmost important (you've got a point for the embedded doc of some modes but once again there are fixed rules for the rebinding). I think that evilified is a good balance between:
- principle of least astonishment
- Vim feeling
- consistency (important given the mutlitude of packages that are thought to work with Emacs style)
- low maintenance burden

For the rest I did not read the OP correctly, I'm sorry for this! I'm all for making it easy to opt-out from evilification.

**Comment by syl20bnr** [comment]
> we should also offer navigation style of evilified and stock emacs

It would be so confusing for the end user.

I guess an explicit variable is the best we can do. @bmag's proposal for `dotspacemacs-non-evilified-modes` looks fine to me and is easy to implement.

**Comment by mbertheau** [comment]
Hmm, @justbur, I added these lines to `dotspacemacs/user-config`
```elisp
  (add-to-list 'evil-emacs-state-modes 'magit-status-mode)
  (add-to-list 'evil-emacs-state-modes 'magit-log-mode)
```

And in magit log <kbd>l</kbd> still moves the point to the right, and only <kbd>L</kbd> triggers the magit log popup. Am I doing something wrong?

**Comment by justbur** [comment]
It's possible the evilify stuff is overriding that setting for you. I don't
know enough about how it works to say right now. I think I did my test
after emacs had fully loaded so it wasn't equivalent to what you did here.

**Comment by mbertheau** [comment]
I tried @justbur's PR and it works nicely. However it requires the user to know which modes provide their own alphabetical keybindings. Can and should we maintain a list of such modes in spacemacs that can be used as a default value for `dotspacemacs-evilify-blacklist`?

**Comment by mbertheau** [comment]
Also I wonder if the <kbd>SPC</kbd> prefix should be retained even for blacklisted modes. It's the universal way out and shouldn't be missing from any mode in spacemacs IMO. I'm not sure how evilifaction works with regard to <kbd>SPC</kbd>, but maybe the mode-specific <kbd>SPC</kbd> binding - if there is one - can be put on <kbd>, SPC</kbd> and <kbd>SPC m SPC</kbd> respectively.

**Comment by syl20bnr** [comment]
> However it requires the user to know which modes provide their own alphabetical keybindings. Can and should we maintain a list of such modes in spacemacs that can be used as a default value for dotspacemacs-evilify-blacklist?

It is OK to me, no matter what we do you'll have to know the mode where you want to opt-out. Now magit has a lot of modes, that's unfortunate. I think we should try another path and drop the idea of the black list.

> Also I wonder if the SPC prefix should be retained even for blacklisted modes. It's the universal way out and shouldn't be missing from any mode in spacemacs IMO. I'm not sure how evilifaction works with regard to SPC, but maybe the mode-specific SPC binding - if there is one - can be put on , SPC and SPC m SPC respectively.

I'm not in favor of introducing special rules that moves the leaders.

**Comment by syl20bnr** [comment]
Let's just do a global opt-out via `dotspacemacs-evilified-buffers` it is simpler and easier to implement. I don't want anymore to allow mixing evilified and raw emacs buffers but I'm ok to provide a global switch.

**Comment by mbertheau** [comment]
Hmm, that's a good point. I now realize that I don't really know what evilification does for modes other than the magit modes. (I don't use dired or gnus or mu4e.) So I don't know if there are any modes that I wouldn't want to be on `dotspacemacs-evilify-blacklist`.

I still think the leader <kbd>SPC</kbd> should be available everywhere. Maybe instead of `dotspacemacs-evilify-buffers` we can have a boolean `dotspacemacs-evilify-full` (or a better name) which if `false` only takes care that the leader is available, as opposed to doing full evilification as documented in https://github.com/syl20bnr/spacemacs/blob/develop/doc/CONVENTIONS.org#evilify-buffers? Or we have `dotspacemacs-evilify-mode` (or a better name), that can be `'none` for no evilification at all, `'leader-only` or `'full`.

**Comment by justbur** [comment]
@mbertheau just a comment. All that pr does is start you in emacs state in
those buffers. The M-m leader and M-RET major mode leader should always be
available in emacs state.

**Comment by syl20bnr** [comment]
What you want is some kind of alternate hybrid style :-) It is hard to support all the possible variants. My principle concern is to keep things simple and starting to mix SPC leader availability in Emacs state is too complicated for my taste and break the fact that Emacs state is pure Emacs :-( 

There is one hope in the user configuration space though! Explore the customization of evil-evilified-state-map to your liking, this way you tailor your own idea of evilification and you don't even need to change the default state.

**Comment by syl20bnr** [comment]
Altering evil-evilified-state-map in dotspacemacs/user-init should work. For instance clear the keymap.

**Comment by justbur** [comment]
I think the blacklist makes things too complicated now that I think about it. For me the simplest thing to do would be to be able to disable evilification in .spacemacs completely. Then you just know that it won't ever come into play. 

@mbertheau Having `SPC` available in emacs state is tricky, because the mode might expect you to be able to enter text. 

@syl20bnr Here's a crazy idea. If people disable evilification, make hybrid their default state in these buffers. That way escaping into normal mode works and you can be comfortable with your muscle memory for `SPC`

Also, one more thought I had is that there might be too many states now for a color-based indication to be obvious enough. I think it might be worth exploring a label somewhere or some more obvious method of signalling to the user what state they are in.

**Comment by syl20bnr** [comment]
> Here's a crazy idea. If people disable evilification, make hybrid their default state in these buffers. That way escaping into normal mode works and you can be comfortable with your muscle memory for SPC

Might work with the addition of opt-in evilification with something like `dotspacemacs-evilified-buffers-in-hybrid`. But I don't think @mbertheau want the hybrid insert state so we are still goind round in circle :-) And I don't think it is consistent with the definition of hybrid style which uses normal state as default and so uses hjkl navigation.

The OP is very specific and it seems that the easiest way to get the requested behavior is just to clear the map of `evil-evilified-state` in `dotspacemacs/user-init` so we keep this is user space.

**Comment by syl20bnr** [comment]
I plan a rewrite of evilification map during the week for the 0.104.0, the current implementation is naive and too complicated, but there won't be anything new to support this use case. Please let keep it in user space, we can help @mbertheau to achieve the requested result and document it somewhere (maybe in @mbertheau FAQ).

**Comment by mbertheau** [comment]
Ok, seems I first have to understand what bare evil does to magit mode, what spacemacs evilification adds and how hybrid and emacs mode play into that before I can contribute something constructive to this issue :)

**Comment by justbur** [comment]
@mbertheau All evil mode does for the most part is to intercept your keypresses and translate them into commands that are meant to emulate vim's behavior. How and when it intercepts the keypresses depends on what state it is in. 

"Emacs state" is almost equivalent to disabling evil mode entirely. It doesn't intercept any keypresses by default in this state. That's why putting evil in emacs state while you're in a magit buffer makes all of the magit key bindings work. 

Hybrid is almost emacs state, but it sets up the necessary functionality to play well with normal state. So the idea is that in hybrid state none of your keypresses are intercepted except the ones related to escaping back into normal state. 

Evilification is a way of keeping the essential vim key bindings (hjkl), but also making the default magit bindings available, so what it does is to move the magit bindings out of the way according to a set of rules to make room for hjkl.

I believe bare evil just puts the magit buffers in emacs state.

**Comment by justbur** [comment]
@syl20bnr I tried clearing the keymap as you suggested, but it left me with a weird set of key bindings in magit-status-mode. I added another commit in #3043 to play with an option to completely disable entering evilified state. Essentially, it just mimics what holy-mode does.

**Comment by mbertheau** [comment]
I tried `dotspacemacs-evilify-blacklist` from #3043 for a while and I found it really irritating that <kbd>SPC</kbd> doesn't work, which is the universal way out (for example to another buffer). I realize that that probably means that in a final solution we could still deviate from on-screen docs for <kbd>SPC</kbd>.

Will try `dotspacemacs-dont-evilify` now.

**Comment by mbertheau** [comment]
@justbur Ok, so evil manages evil states on top of the current major mode for the buffer. They start in normal state where `hjkl`, `:`, `/`, `n`, `p` and so on all work. `i` and others switch to insert state. I don't really do anything in insert mode except for typing characters that appear in the buffer and escaping to normal mode, so I can't tell whether there's a difference between insert state and the major mode without evil. Evidently there must be because hybrid exists. What is the difference between hybrid and vim mode for spacemacs?

Now maybe I understand evilification: in normal state evil grabs maybe all but certainly a lot of the key bindings from the major mode. With hybrid and to some extent with spacemacs vim mode you have all the major mode bindings in insert state available (except for bindings that escape to normal state).

Since it is awkward to have an insert and a normal mode for major modes that are not about editing text, spacemacs added evilification. It is a kind of light normal state where most of the major mode bindings work except for `hjklnp/` and so on as documented. There's no insert state in evilified modes.

Is this correct?

**Comment by mbertheau** [comment]
With my current understanding I believe what I want is a light evilification that only adds the <kbd>SPC</kbd> leader (if spacemacs mode is vim) and doesn't overwrite `nphjkl` or other keys.

**Comment by mbertheau** [comment]
<kbd>SPC</kbd> doesn't work with `dont-evilify` either, but my understanding improved greatly :)

I'm running with light evilification now https://github.com/mbertheau/spacemacs/commit/11e9c5675a5d931559cf7b550cecf936917b2f9e and will report any issues.

**Comment by justbur** [comment]
@mbertheau 
>so I can't tell whether there's a difference between insert state and the major mode without evil. Evidently there must be because hybrid exists. What is the difference between hybrid and vim mode for spacemacs?

[Here are the key bindings in insert state, which also exist in vim](https://bitbucket.org/lyro/evil/src/b22e7e7b2526c3f1ae274dbf13b6e6d3ea7a8cd3/evil-maps.el?at=default&fileviewer=file-view-default#evil-maps.el-353)

>Since it is awkward to have an insert and a normal mode for major modes that are not about editing text, spacemacs added evilification. It is a kind of light normal state where most of the major mode bindings work except for hjklnp/ and so on as documented. There's no insert state in evilified modes.
>Is this correct?

Mostly. Think of it this way. magit sets up an extensive set of key bindings for using it. Putting evil in normal state in a magit buffer wouldn't work because the normal state commands in evil mostly have nothing to do with magit. In emacs state, the full magit keymap is available, but it uses keys that are unfamiliar to the hjkl crowd, so evilification uses rules to alter the magit keymap on the fly, making j go down one line, moving whatever j did up to J, moving whatever J did somewhere else, etc. 

It wouldn't make much sense to be in insert state in most magit buffers, because the few insert state bindings there are could clash with the magit bindings.

Basically, if you are comfortable with the default magit keys, you should be in emacs state in magit buffers, but the `SPC` key doesn't get bound in emacs state because it would kind of be hard to type something meaningful in emacs state other places :wink:.

What you could do is bind `SPC` in the magit keymaps and then go into emacs state in those buffers. Does that make sense?

**Comment by mbertheau** [comment]
Hmm, what about modifying evilify for these modes such that keys that are bound in the major mode are not overwritten with evil commands? Or the evil commands are placed on the first free key of the `n` -> `N` -> `C-n` -> `C-N`  sequence. The spacemacs leader key should always be bound to spacemacs-menu though. If the major mode defines a binding for `SPC`, this binding can be moved to `C-SPC` or something else.

**Comment by justbur** [comment]
@mbertheau It sounds like what you want is the magit key bindings with `SPC` available. Is that right? In that case, I would just put `SPC` in `magit-mode-map`. Try this with the dont-evilify option in #3043 
```elisp
(define-key magit-mode-map " " evil-leader--default-map)
```

**Comment by mbertheau** [comment]
@justbur I've been running with https://github.com/mbertheau/spacemacs/commit/11e9c5675a5d931559cf7b550cecf936917b2f9e in the last few days and can report near perfect happiness. I'd just like the command that magit binds to `SPC` to be available. It could maybe be on `C-SPC` or on `SPC SPC` or `SPC m SPC` and `, SPC` respectively. 

I will try to figure out what the exact effective difference between https://github.com/mbertheau/spacemacs/commit/11e9c5675a5d931559cf7b550cecf936917b2f9e and your proposal in the previous comment is.

**Comment by justbur** [comment]
@mbertheau it should not be a big difference at all. However, I appreciate having the sky blue color for emacs state, because it tells me what to expect. Under your strategy you should still see the pale yellow for evilified state, but your evilified state is really mimicking emacs state. 

You can move the `SPC` commands for magit, but instead of worrying about where the space command went in all evilified buffers my suggestion would be to try and get used to the emacs leader key (default `M-m`) in these cases. If you can do that, you don't need to worry about rebinding `SPC` at all.

**Comment by justbur** [comment]
note: magit will no longer be evilified after #3578
